diff --git a/src/main.cpp b/src/main.cpp
index b30210b..bba58ae 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -770,6 +770,7 @@ APICALL EXPORT PLUGIN_DESCRIPTION_INFO PLUGIN_INIT(HANDLE handle) {
     HyprlandAPI::addConfigValueV2(PHANDLE, g_config->sensitivity);
     HyprlandAPI::addConfigValueV2(PHANDLE, g_config->sendCancel);
     HyprlandAPI::addConfigValueV2(PHANDLE, g_config->resizeOnBorder);
+    HyprlandAPI::addConfigValueV2(PHANDLE, g_config->workspaceSwipeEdge);

     static auto P0 = Event::bus()->m_events.config.preReload.listen([&] { onPreConfigReload(); });

